enhancements/authentication/pod-security-admission-enforcement.md

```diff
@@ -0,0 +1,419 @@
+
+- resolve the violating Namespaces in order to be able to `Upgrade` or
+- set the `spec.enforcementMode=Privileged` and solve the violating Namespaces later.
+
+If a user manages several clusters and there are well known violating Namespaces, the `spec.enforcementMode=Privileged` can be set as a precaution.
+
+### Implementation Details
+
+- The `PodSecurityReadinessController` in the `cluster-kube-apiserver-operator` will manage the new API.
+- The [`Config Observer Controller`](https://github.com/openshift/cluster-kube-apiserver-operator/blob/218530fdea4e89b93bc6e136d8b5d8c3beacdd51/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L131) must be updated to watch for the new API's status alongside the `FeatureGate`.
+- The [`PodSecurityAdmissionLabelSynchronizationController`](https://github.com/openshift/cluster-policy-controller/blob/master/pkg/cmd/controller/psalabelsyncer.go#L17-L50) must be updated to watch for the new API's status alongside the `FeatureGate`.
+- If the `FeatureGate` `OpenShiftPodSecurityAdmission` is removed from the current `FeatureSet`, the cluster must revert to its previous state.
+  It serves as a break-glass mechanism.
+
+#### PodSecurityReadinessController
+
+The `PodSecurityReadinessController` will manage the `PSAEnforcementConfig` API.
+It already collects most of the necessary data to determine whether a Namespace would fail enforcement or not to create a [`ClusterFleetEvaluation`](https://github.com/openshift/enhancements/blob/master/dev-guide/cluster-fleet-evaluation.md).
+With the `security.openshift.io/MinimallySufficientPodSecurityStandard`, it will be able to evaluate all Namespaces for failing workloads, if any enforcement would happen.
+With the `security.openshift.io/ValidatedSCCSubjectType`, it can categorize violations more accurately and create a more accurate `ClusterFleetEvaluation`.
+
+#### PodSecurity Configuration
+
+A Config Observer in the `cluster-kube-apiserver-operator` manages the Global Config for the kube-apiserver, adjusting behavior based on the `OpenShiftPodSecurityAdmission` `FeatureGate`.
+It must watch both the `status.enforcementMode` for `Restricted` and the `FeatureGate` `OpenShiftPodSecurityAdmission` to be enabled to make decisions.
+
+#### PodSecurityAdmissionLabelSynchronizationController
+
+The [PodSecurityAdmissionLabelSynchronizationController (PSA label syncer)](https://github.com/openshift/cluster-policy-controller/blob/master/pkg/psalabelsyncer/podsecurity_label_sync_controller.go) must watch the `status.enforcementMode` and the `OpenShiftPodSecurityAdmission` `FeatureGate`.
+If `spec.enforcementMode` is `Restricted` and the `FeatureGate` `OpenShiftPodSecurityAdmission` is enabled, the syncer will set the `pod-security.kubernetes.io/enforce` label on Namespaces that it manages.
+Otherwise, it will refrain from setting that label and remove any enforce labels it owns if existent.
```

Comment:
Are there any other changes that need to be made here?

How should the label syncer handle user-overridden audit/warn labels when deciding whether or not to set the enforce label?

Does any evaluation on the existing workloads in the namespace need to take place to determine if setting the enforce label will break workloads? If setting the enforce label will break workloads in the namespace, how should the controller react?

Reply:
> How should the label syncer handle user-overridden audit/warn labels when deciding whether or not to set the enforce label?

User labels are ignored, see here: https://github.com/openshift/enhancements/pull/1747/files#r1971301549

> Does any evaluation on the existing workloads in the namespace need to take place to determine if setting the enforce label will break workloads? If setting the enforce label will break workloads in the namespace, how should the controller react?

It is not the concern of the controller to identify if the workload would break or not. The concern is to decide what label it thinks is appropriate and set if.
It is the concern of the PodSecurityReadinessController to identify if it is safe to enforce PSA.